src/com/twitter/intellij/pants/bsp/JarMappings.java

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+package com.twitter.intellij.pants.bsp;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.vfs.LocalFileSystem;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import com.intellij.openapi.vfs.newvfs.BulkFileListener;
+import com.intellij.openapi.vfs.newvfs.events.VFileContentChangeEvent;
+import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.reflect.Type;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class JarMappings {
+  private static final String SOURCES_JAR_SUFFIX = "-sources.jar";
+
+  public static JarMappings getInstance(Project project) {
+    return ServiceManager.getService(project, JarMappings.class);
+  }
+
+  public JarMappings(Project project) {
+    this.project = project;
+
+    project.getMessageBus().connect().subscribe(VirtualFileManager.VFS_CHANGES, new BulkFileListener() {
+      @Override
+      public void after(@NotNull List<? extends VFileEvent> events) {
+        events.forEach(event -> {
+          if (event instanceof VFileContentChangeEvent &&
+              event.getFile() != null &&
+              event.getFile().equals(librariesFile())) {
+            librariesFileIsUpToDate = false;
+          }
+        });
+      }
+    });
+  }
+
+  private final Project project;
+
+  private Map<String, String> libraryJarToLibrarySourceJar = new HashMap<>();
+  private boolean librariesFileIsUpToDate = false;
+
+  private synchronized void ensureInitialized() {
+    try {
+      if (!librariesFileIsUpToDate) {
+        VirtualFile file = librariesFile();
+        String content = new String(file.contentsToByteArray());
+        Type mapType = new TypeToken<Map<String, String>>() {}.getType();
+        libraryJarToLibrarySourceJar = new Gson().fromJson(content, mapType);
+        librariesFileIsUpToDate = true;
+      }
+    }
+    catch (Exception e) {
+      librariesFileIsUpToDate = false;
```

Comment:
I mostly expect it to fail when file is not there (could be also on bad json though), then we will end up with empty map or the previous version which should be fine to proceed. This will make sure we will just retry on the next call. But logging would be helpful in case of debugging. I added logging.